### ISSUE TITLE
Crash on current bleed in the Move activity

### ISSUE THREAD

**Issue by obrakmann** [OpenRA/OpenRA#7630]
```
Available mods:
	cnc: Tiberian Dawn (git-762c0e1)
	d2k: Dune 2000 (git-762c0e1)
	modchooser: Mod Chooser (git-762c0e1)
	ra: Red Alert (git-762c0e1)
	ts: Tiberian Sun ({DEV_VERSION})
Loading mod: ra
Game started
Exception of type `System.IndexOutOfRangeException`: Array index is out of range.
  at OpenRA.Map.OffsetOfSubCell (SubCell subCell) [0x00001] in /home/oliver/devel/openra/git/OpenRA.Game/Map/Map.cs:156 
  at OpenRA.Mods.Common.Activities.Move+MoveFirstHalf.OnComplete (OpenRA.Actor self, OpenRA.Mods.Common.Traits.Mobile mobile, OpenRA.Mods.Common.Activities.Move parent) [0x0007a] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Activities/Move/Move.cs:381 
  at OpenRA.Mods.Common.Activities.Move+MovePart.InnerTick (OpenRA.Actor self, OpenRA.Mods.Common.Traits.Mobile mobile) [0x0003c] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Activities/Move/Move.cs:331 
  at OpenRA.Mods.Common.Activities.Move+MovePart.Tick (OpenRA.Actor self) [0x00015] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Activities/Move/Move.cs:315 
  at OpenRA.Traits.Util.RunActivity (OpenRA.Actor self, OpenRA.Activities.Activity act) [0x00017] in /home/oliver/devel/openra/git/OpenRA.Game/Traits/Util.cs:83 
  at OpenRA.Actor.Tick () [0x00010] in /home/oliver/devel/openra/git/OpenRA.Game/Actor.cs:136 
  at OpenRA.World.Tick () [0x000de] in /home/oliver/devel/openra/git/OpenRA.Game/World.cs:304 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x001bf] in /home/oliver/devel/openra/git/OpenRA.Game/Game.cs:458 
  at OpenRA.Game.LogicTick () [0x0004e] in /home/oliver/devel/openra/git/OpenRA.Game/Game.cs:482 
  at OpenRA.Game.Loop () [0x000d4] in /home/oliver/devel/openra/git/OpenRA.Game/Game.cs:598 
  at OpenRA.Game.Run () [0x00042] in /home/oliver/devel/openra/git/OpenRA.Game/Game.cs:638 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in /home/oliver/devel/openra/git/OpenRA.Game/Support/Program.cs:111 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /home/oliver/devel/openra/git/OpenRA.Game/Support/Program.cs:39 
```

The debugger tells me that the out-of-range value is `SubCell.Invalid` which is passed to `OffsetOfSubCell()`

Got a bit further:
`MoveFirstHalf.OnComplete()`: `var nextCell = parent.PopPath(self);`
`Move.PopPath()`: `var subCell = mobile.GetAvailableSubCell(nextCell, SubCell.Any, ignoredActor);`
`Mobile.GetAvailableSubCell`: `return Info.GetAvailableSubCell(...)`
`MobileInfo.GetAvailableSubCell`:
```csharp
			if (MovementCostForCell(world, cell) == int.MaxValue)
				return SubCell.Invalid;
```
So that's how SubCell.Invalid ends up as an argument to `OffsetOfSubCell`.

I don't know my way around these parts though, so I'm not sure if we should just fix it with the patch below, or do something else in case this is just part of a bigger problem. In particular, I wonder why `PopPath` would return a position with an invalid subcell (ie. where the movement cost is `int.MaxValue`) in the first place.

```diff
diff --git a/OpenRA.Mods.Common/Activities/Move/Move.cs b/OpenRA.Mods.Common/Activities/Move/Move.cs
index 5ff3ee0..0f6c4c5 100644
--- a/OpenRA.Mods.Common/Activities/Move/Move.cs
+++ b/OpenRA.Mods.Common/Activities/Move/Move.cs
@@ -378,7 +378,9 @@ protected override MovePart OnComplete(Actor self, Mobile mobile, Move parent)
                                {
                                        if (IsTurn(mobile, nextCell.Value.First))
                                        {
-                                               var nextSubcellOffset = self.World.Map.OffsetOfSubCell(nextCell.Value.Second);
+                                               var nextSubCell = nextCell.Value.Second;
+                                               var nextSubcellOffset = nextSubCell != SubCell.Invalid ? self.World.Map.OffsetOfSubCell(nextSubCell) : WVec.Zero;
                                                var ret = new MoveFirstHalf(
                                                        Move,
                                                        Util.BetweenCells(self.World, mobile.FromCell, mobile.ToCell) + (fromSubcellOffset + toSubcellOffset) / 2,

```

**Comment by obrakmann** [comment]
@Rydra, can you shed any light on this? ^

**Comment by Rydra** [comment]
uhmm @obrakmann  do you have any steps to reproduce this?

**Comment by Rydra** [comment]
Also @obrakmann I invite you to try #7618 and see if it fixes this error.

**Comment by obrakmann** [comment]
No, I don't have any steps to reproduce. I've also only seen it in one game so far. But I have [a replay](https://seacloud.cc/f/439a4b4012/). You need [this map](https://seacloud.cc/f/993fc07256/) to view it.

**Comment by Rydra** [comment]
Definitely, my PR should fix that error. If it happens again I'll take a closer look at it.

**Comment by RoosterDragon** [comment]
This should be fixed with #7618 being merged.

**Comment by obrakmann** [comment]
Alright, I'll close this then.
